book/src/dev/rfcs/0004-asynchronous-script-verification.md

```diff
@@ -82,18 +97,162 @@ parallelism.
+/// MUST have no transparent outputs (i.e.tx_out_count MUST be 0).
+enum SpendRestriction {
+    /// The UTXO is spent in a transaction with transparent outputs
+    SomeTransparentOutputs,
+    /// The UTXO is spent in a transaction with all shielded outputs
+    AllShieldedOutputs,
+}
+
+enum Response::Utxo(transparent::Output)
+```
+
+As described above, the request name is intended to indicate the request's behavior.
+The request does not resolve until:
+- the state layer learns of a UTXO described by the request, and
+- the output is spendable at `height` with `spend_restriction`.
+
+We also add a coinbase flag and height to `transparent::Output`s that we look up in
+the state, or get from newly commited blocks:
+```rust
+enum TransparentOutputOrigin {
+    Coinbase {
+        output: transparent::Output,
+        height: Height,
+    },
+    NonCoinbase {
+        output: transparent::Output,
+    },
+}
+```
+
+The coinbase flag and height can be derived from the `tx_index` and `height`:
+- looked up from the `OutPoint.hash` in `tx_by_hash`, or
+- retrieved from the newly arrived block.
+
+There is exactly one coinbase transaction per block, and it must have an index of `0`.
+
+## Transparent coinbase consensus rules
+[transparent-coinbase-consensus-rules]: #transparent-coinbase-consensus-rules
+
+Specifically, if the UTXO is a transparent coinbase output,
+the service is not required to return a response if:
+- `spend_height` is less than `MIN_TRANSPARENT_COINBASE_MATURITY` (100) blocks after the `Coinbase.height`, or
+- `spend_restriction` is `SomeTransparentOutputs`.
+
+This implements the following consensus rules:
+
+> A transaction MUST NOT spend a transparent output of a coinbase transaction
+> from a block less than 100 blocks prior to the spend.
+>
+> Note that transparent outputs of coinbase transactions include Founders’ Reward
+> outputs and transparent funding stream outputs.
+
+> A transaction with one or more transparent inputs from coinbase transactions
+> MUST have no transparent outputs (i.e.tx_out_count MUST be 0).
+>
+> Inputs from coinbase transactions include Founders’ Reward outputs and funding
+> stream outputs.
+
+https://zips.z.cash/protocol/protocol.pdf#txnencodingandconsensus
+
+## Parallel coinbase checks
+[parallel-coinbase-checks]: #parallel-coinbase-checks
+
+We can perform these coinbase checks asynchronously, in the presence of multiple chain forks,
+as long as the following conditions hold.
+
+We don't mistakenly accept or reject spends to the transparent pool:
+- across all chains, the set of coinbase transaction hashes is disjoint from
+  the set of non-coinbase transaction hashes.
+
+We don't mistakenly accept or reject mature spends:
+- across all chains, the set of coinbase transaction hashes is disjoint from
+  the set of non-coinbase transaction hashes; and
+- across all chains, duplicate coinbase transaction hashes can only occur at
+  exactly the same height.
```

Comment:
It would be useful to provide a rationale for the conditions. Why are these conditions needed? Why are them enough?

That's implicit, and I think I was able to deduce it (e.g. for the first one, if there is a repeated hash then we could mistakenly think that the coinbase input of a transaction it's not a coinbase input, and allow it to be spent to a unshielded address?), but it would be easier for the reader to make it explicit.

However that's not critical so feel free to merge.

Reply:
I think a rationale is a great idea, and I'll do it when I do the revisions.

Reply:
I added a justification in 25dfa1fa